### PR TITLE
setting tests

### LINKED ISSUE #2: automate the validation of the PR
## Context

Check if the submitted plugin follows the current standard 

## User flow

run some unit tests in the CI that runs for each push, the pr will be accepted only if the runs succed

## How to

- create a unit testing file that checks if the json fullfill all the requirements
- call it in the CI

## Technical details

I will be using python with pytest and unittest

### PR DESCRIPTION
**Pull request by Pierregvx**
added basic tests to check if all the elements in `plugins.json` are following the right structure & contains a  zip file

### CODE REVIEW

**Comment by Pierregvx** [comment]
I tried to do an invalid pr, [the run failed as expected](https://github.com/massalabs/thyra-plugin-store/actions/runs/4669793422/jobs/8268716164) but it looks like the branch still can  be merged

**Comment by Pierregvx** [comment]
the tests are failing because I didn't update the plugin file

**Comment by Pierregvx** [comment]
@peterjah could you enable anyone to merge a pr if the run is green, but only admins if it is red in the repo settings ? cc @LucasF243

File: test/plugin.test.ts [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+  test("if one plugin has been added, all the other are exactly the same", () => {
+    if (plugins.length > lastVersion.length) {
+      for (const data of lastVersion) {
+        expect(plugins).toContainEqual(data);
+      }
+    }
+  });
+  test("one plugin is updated if there is an update", () => {
+    if (plugins.length == lastVersion.length) {
+      // get the number of plucin where the version has been updated in one line
```
Comment (0xMazout) [inline test/plugin.test.ts]:
```suggestion
      // get the number of plugin where the version has been updated in one line
```

File: plugins.json [hunk 2]
```diff
@@ -1,26 +1,27 @@
 [
```
Comment (Pierregvx) [inline plugins.json:1]: I changed this for the tests

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1,3 @@
+node_modules/
+yarn-error.log
+yarn.lock
```
Comment (peterjah) [inline .gitignore]: the `yarn.lock` should not be ignored. just commit it !

File: package.json [hunk 4]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "my-js-test-project",
```
Comment (peterjah) [inline package.json]: choose a name here :p

File: .github/workflows/plugin-CI.yml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+    - name: Run tests
+      run: |
+        yarn test |& tee /tmp/test.log
+        if grep -q "console.warn" /tmp/test.log; then
+          echo "::warning::Tests produced warnings!"
+        fi
+        if grep -q "Test Suites:.* failed" /tmp/test.log; then
+          exit 1
+        fi
```
Comment (peterjah) [inline .github/workflows/plugin-CI.yml]: i don't really see the point of this. `yarn test` should return an exit code > 0 if some test are failing.
and if you need the logs, you can just go in github action page

**Review by peterjah — CHANGES_REQUESTED** [review]

File: .github/workflows/plugin-CI.yml [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  pull_request:
+    branches:
+      - main
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v2
```
Comment (Thykof) [inline .github/workflows/plugin-CI.yml]:
```suggestion
    - uses: actions/checkout@v3
```

File: test/plugin.test.ts [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+        (plugin, index) =>
+          plugin.version != lastVersion[index].version &&
+          plugin.name == lastVersion[index].name
+      );
+      const updatedPluginsAmount = updatedPlugins.length;
+      if (updatedPluginsAmount === 0) {
+        throw new Error("No plugin has been updated");
+      } else if (updatedPluginsAmount > 1) {
+        throw new Error("More than one plugin have been updated");
+      }
```
Comment (mazmassa) [inline test/plugin.test.ts]: I am having a hard time to got the point here. What's the aim in throwing errors in the test!? 🤔

File: .github/workflows/plugin-CI.yml [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+        node-version: 16.x
+    - name: Install Yarn
+      run: npm install -g yarn
+      
+    - name: Install dependencies
+      run: yarn install --frozen-lockfile
+
+    - name: Run tests
+      run: |
+        yarn test |& tee /tmp/test.log
```
Comment (Thykof) [inline .github/workflows/plugin-CI.yml]: no-blocking, why do you want to save the output of the yarn test command onto a file? how will you read the content of that file in a context of github runners?

File: test/plugin.test.ts (hunk 7 above)
Comment (Pierregvx) [reply test/plugin.test.ts]: It was to return a certain error message, but you're right I should change that

File: test/plugin.test.ts [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+} from "./utils";
+
+describe("TestFormat", () => {
+  let plugins: any[];
+  let lastVersion: any[];
+
+  beforeAll(async () => {
+    plugins = await jsonfile.readFile("plugins.json");
+    lastVersion = await fetch(
+      "https://raw.githubusercontent.com/massalabs/thyra-plugin-store/main/plugins.json"
```
Comment (mazmassa) [inline test/plugin.test.ts]: WDYT to place that in a constant on the top!?

**Review by Thykof — CHANGES_REQUESTED** [review]
my partial review

File: .github/workflows/plugin-CI.yml (hunk 8 above)
Comment (Pierregvx) [reply .github/workflows/plugin-CI.yml]: I just removed it

File: test/plugin.test.ts (hunk 9 above)
Comment (Pierregvx) [reply test/plugin.test.ts]: yep you're right

File: test/plugin.test.ts [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+
+  let pluginAssets: { [key: string]: PluginAsset } = {};
+
+  beforeAll(async () => {
+    plugins = await jsonfile.readFile("plugins.json");
+
+    // Fetch the last version of the plugins file from GitHub
+    lastVersion = await fetch(
+      "https://raw.githubusercontent.com/massalabs/thyra-plugin-store/main/plugins.json"
+    ).then((res) => res.json());
```
Comment (mazmassa) [inline test/plugin.test.ts]: Don't you think we should have only one beforeAll by file tests!? Like that we all know that the setup for the down tests is done there. Is it possible split these tests in two files and place the beforeAll only once by file!?

WDYT!?

File: test/utils/maths.ts [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+  version2: string
+): boolean => {
+  const version1Array = version1.split(".");
+  const version2Array = version2.split(".");
+  for (let i = 0; i < version1Array.length; i++) {
+    if (parseInt(version1Array[i]) > parseInt(version2Array[i])) {
+      return true;
+    }
+  }
+  return false;
```
Comment (Thykof) [inline test/utils/maths.ts]: no need to reinvent the wheel :) https://www.npmjs.com/package/semver you can use semver.gt

even better you don't need this function, nor this math.ts file because you can directly semver.gt in the code

File: test/plugin.test.ts (hunk 10 above)
Comment (mazmassa) [reply test/plugin.test.ts]: It would allow you to send split MR with less code for review and tests segmented. WDYT!?

File: test/utils/zip.ts [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+  return entries.map((entry) => entry.name);
+}
+
+/**
+ * Calculates the checksum of a file
+ * @param {string} url - The URL of the file to calculate the checksum for
+ * @returns {Promise<string>} - A promise that resolves with the checksum of the file
+ */
+
+export async function calculateChecksum(url: string): Promise<string> {
```
Comment (Thykof) [inline test/utils/zip.ts]: this function seems not to be used elsewhere than in the file, no need to export

```suggestion
async function calculateChecksum(url: string): Promise<string> {
```

File: test/plugin.test.ts [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+            const files = await getZipFileList(asset.url);
+            const checksum = asset.checksum;
+            return [key, { ...asset, files, checksum }] as [
+              string,
+              PluginAsset
+            ];
+          }
+        )
+      ).then(Object.fromEntries));
+  }, 30000);
```
Comment (mazmassa) [inline test/plugin.test.ts]: It would be hard to follow that in future... Is there a way to reduce the complexity here!?

Extracting to some named functions, less logic etc!? 🤔

**Review by Thykof — CHANGES_REQUESTED** [review]

File: test/plugin.test.ts [hunk 14]
```diff
@@ -139,7 +137,7 @@ describe("tests on updated or new plugin", () => {
   test("plugin's logo is a valid image", async () => {
     const dimLogo = await getImageDimensions(plugin.logo);
     expect(!!dimLogo).not.toBeNull();
-    expect(dimLogo!.width!<40).toBe(true);
+    expect(dimLogo!.width! < 40).toBe(true);
```
Comment (Thykof) [inline test/plugin.test.ts]: don't hesitate to use a code formatting tool such as prettier ;) (github.com/massalabs/prettier-config-as)

File: test/plugin.test.ts [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+  }, 30000);
+  test("plugin's logo is a valid image", async () => {
+    const dimLogo = await getImageDimensions(plugin.logo);
+    expect(!!dimLogo).not.toBeNull();
+    expect(dimLogo!.width!<40).toBe(true);
+  });
+  platforms.forEach(({ name, pattern }) => {
+    // tests triggered if there is a new or updated plugin
+
+    test(`${name} zip contains a ${name} executable`, async () => {
```
Comment (mazmassa) [inline test/plugin.test.ts]: I wouldn't use loop in tests. Tests allow people to understand what is happening. It's a way to learn the code. If we add more logic we could be easily lost.

Is there an special reason for that I am missing!?

Otherwise you should cover each case explicitly. It will help to understand what we do in each case. Let me know if I am not clear. 🙂

**Review by Thykof — CHANGES_REQUESTED** [review]

File: test/plugin.test.ts (hunk 13 above)
Comment (Pierregvx) [reply test/plugin.test.ts]: True, I'll try to simplify this

File: test/utils/maths.ts (hunk 11 above)
Comment (Pierregvx) [reply test/utils/maths.ts]: thanks ! I'll use it

File: test/plugin.test.ts [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+
+import { structure, imageFormats } from "./standard";
+import {
+  checkFileChecksum,
+  getZipFileList,
+  getImageDimensions,
+  isVersionBiggerThan,
+} from "./utils";
+
+const storeListURL = "https://raw.githubusercontent.com/massalabs/thyra-plugin-store/main/plugins.json"
```
Comment (mazmassa) [inline test/plugin.test.ts]: const URL
const LIST_URL 

Have a look here:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const#examples

File: test/plugin.test.ts (hunk 10 above)
Comment (Pierregvx) [reply test/plugin.test.ts]: yes It would give more visibility

File: test/plugin.test.ts [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+      }
+    }
+  });
+  test("one plugin is updated if there is an update", () => {
+    if (plugins.length == lastVersion.length) {
+      // get the number of plugin where the version has been updated in one line
+      const updatedPlugins = plugins.filter(
+        (plugin, index) =>
+          plugin.version != lastVersion[index].version &&
+          plugin.name == lastVersion[index].name
```
Comment (mazmassa) [inline test/plugin.test.ts]: Instead of that, we could use for example:
```
let isSameVersion = plugin.version == lastVersion[index].version (note I change the comparison)
let hasSameName = plugin.name == lastVersion[index].name
```
then you can use:

```
! isSameVersion && hasSameName ... 
```

Isn't more readable, explicit and easy to understand. WDYT!?

File: test/plugin.test.ts [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+      );
+      const updatedPluginsAmount = updatedPlugins.length;
+      expect(updatedPluginsAmount).toBe(1);
+      if (updatedPluginsAmount == 1) {
+        // get the plugin that has been updated in the last version
+        const lastVersionUpdatedPlugin = lastVersion.filter(
+          (plugin, index) => plugin.version != plugins[index].version
+        )[0];
```
Comment (mazmassa) [inline test/plugin.test.ts]: Could you, please, double check of all your filters!? I am seeing on each test. I have no context but I have the feeling they are repeated. 

About array position `[0]` if we could be more declarative would be better. It can be tricky. I guess in that case the`slice()` method does the job. Isn't it!? 🤔

File: plugins.json [hunk 19]
```diff
@@ -1,26 +1,27 @@
...
             },
             "macos-amd64": {
-                "url": "https://github.com/massalabs/thyra-node-manager-plugin/releases/download/v0.0.8/thyra-plugin-node-manager_darwin-amd64.zip",
-                "checksum": "247655ea2985c488c678e195699684d7"
+                "url": "https://github.com/massalabs/thyra-node-manager-plugin/releases/download/v0.0.10/thyra-plugin-node-manager_darwin-amd64.zip",
+                "checksum": "a179319ee6d8e5caae3f945989a8135a"
             }
         },
-        "version": "0.0.8",
+        "version": "0.0.18",
```
Comment (Thykof) [inline plugins.json]: why not the same version as the release, v0.0.10?

**Review by Thykof — COMMENTED** [review]
no blocking, I would have done it differently, I'm just sharing with you: create a function that validate the json store, add unit tests for this function with one invalid json store for every conditions you want to check. in the CI, run this function against the json store and return exit code 1 if not valid. This way, you tests all the code with datasets for all possibilities (all failing cases).

File: plugins.json (hunk 19 above)
Comment (Pierregvx) [reply plugins.json]: this was just for the tests

Comment (Pierregvx) [reply plugins.json]: > no blocking, I would have done it differently, I'm just sharing with you: create a function that validate the json store, add unit tests for this function with one invalid json store for every conditions you want to check. in the CI, run this function against the json store and return exit code 1 if not valid. This way, you tests all the code with datasets for all possibilities (all failing cases).

I think that you are right it makes more sense. Do you agree @mazmassa ?

**Comment by Pierregvx** [comment]
> no blocking, I would have done it differently, I'm just sharing with you: create a function that validate the json store, add unit tests for this function with one invalid json store for every conditions you want to check. in the CI, run this function against the json store and return exit code 1 if not valid. This way, you tests all the code with datasets for all possibilities (all failing cases).

I think that you are right it makes more sense. Do you agree @mazmassa ?

**Review by peterjah — APPROVED** [review]

File: plugins.json [hunk 20]
```diff
@@ -1,26 +1,26 @@
...
-            "macos-amd64": {
-                "url": "https://github.com/massalabs/thyra-node-manager-plugin/releases/download/v0.0.8/thyra-plugin-node-manager_darwin-amd64.zip",
-                "checksum": "247655ea2985c488c678e195699684d7"
-            }
-        },
-        "version": "0.0.8",
-        "url": "https://github.com/massalabs/thyra-node-manager-plugin"
+  {
+    "name": "Node Manager",
+    "description": "Install and manage Massa nodes from Thyra",
```
Comment (mazmassa) [inline plugins.json]: There's no more Thyra! From now on MassaStation

File: plugins.json [hunk 21]
```diff
@@ -1,26 +1,26 @@
...
+    "assets": {
+      "windows": {
+        "url": "https://github.com/massalabs/thyra-node-manager-plugin/releases/download/v0.0.8/thyra-plugin-node-manager_windows-amd64.zip",
```
Comment (mazmassa) [inline plugins.json]: Just ensure that we wont end up with massa-station paths...

File: tests/utils/standard-requirements.ts [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+  definitions: {
+    asset: {
+      type: "object",
+      properties: { url: { type: "string" }, checksum: { type: "string" } },
+      required: ["url", "checksum"],
+    },
+  },
+};
+
+export const IMAGE_FORMATS = [
```
Comment (mazmassa) [inline tests/utils/standard-requirements.ts:29]: 👏

File: tests/validate-store-update.ts [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+    "macos-amd64": /_darwin-amd64$/,
+    "macos-arm64": /_darwin-arm64$/,
+
+}
+function areAllFilesInZipValid(files: Array<string>, pattern: RegExp) {
+    return files.every(
+        // all files must be either a binary either a manifest either an image
+        (file) =>
+            (!!file.match(pattern) !== // file is a binary
+            IMAGE_FORMATS.some((format) => file.endsWith("." + format))) !== // file is an image
```
Comment (mazmassa) [inline tests/validate-store-update.ts]:
```js
IMAGE_FORMATS.some( format => file.endsWith(`.${format}`)))
```

File: tests/validate-store-update.ts [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+    if (!allPluginsHaveDifferentNames) {
+        throw new Error("Error: Plugin name is duplicated");
+    }
+
+    const onlyOneUpdateAndNoNew = newPlugins.length == 0 && updatedPlugins.length == 1
+    const onlyOneNewAndNoUpdate = newPlugins.length == 1 && updatedPlugins.length == 0
+    const onlyOneChange = onlyOneNewAndNoUpdate || onlyOneUpdateAndNoNew
+
+    if (!onlyOneChange) {
+        throw new Error("There should be either one new plugin or one updated plugin");
```
Comment (mazmassa) [inline tests/validate-store-update.ts]: 👏

**Review by mazmassa — APPROVED** [review]
LGTM Pierre! Thanks for the changes!  👏 

I would just format with prettier. You have some diffs in spaces and names. 

Try run : 
`yarn prettier --write .`

File: plugins.json (hunk 21 above)
Comment (Pierregvx) [reply plugins.json]: the `plugins.json` will be modified if the plugin is updated. I didn't make changes here but I'll take your comment into consideration for #6

**Review by Thykof — APPROVED** [review]
